File: TSUUIDKit/Sources/TSUUIDKit/Sync/SyncBundle.swift
import Foundation

/// Wire schema version. Bumped when ``SyncRow`` field layout changes in a
/// way older readers cannot safely ignore. Readers MUST refuse rows with
/// ``schema_version > syncSchemaVersion`` (forward-compat drift guard).
public let syncSchemaVersion: Int = 1

/// Cross-device row sync bundle format.
///
/// Each bundle file is JSON Lines (one row per line). Each row carries the
/// full metadata needed to upsert a vector on the receiving side. This is
/// deliberately separate from the raw `SparseDelta` bundle in ``SyncEngine``:
/// SparseDelta has no `path` field (it was designed for intra-document
/// version diffing), so cross-device row sync needs its own envelope.
///
/// Wire format parity with Python ``tsuuid.sync_bundle`` module.
public struct SyncRow: Codable, Sendable, Equatable {
    public let path: String
    public let title: String
    public let domain: String
    public let vec_b64: String   // base64 of Vector768 float16 bytes (1536 B)
    public let version: Int
    public let encoded_at: String?
    /// Wire schema version for this row. Defaults to ``syncSchemaVersion``.
    /// Optional on decode so pre-schema_version bundles continue to parse
    /// (treated as schema_version = 1).
    public let schema_version: Int

    public init(path: String, title: String, domain: String,
                vec_b64: String, version: Int = 1,
                encoded_at: String? = nil,
                schema_version: Int = syncSchemaVersion) {
        self.path = path
        self.title = title
        self.domain = domain
        self.vec_b64 = vec_b64
        self.version = version
        self.encoded_at = encoded_at
        self.schema_version = schema_version
    }

    public init(path: String, title: String, domain: String,
                vec: Vector768, version: Int = 1,
                encodedAt: Date = Date(),
                schema_version: Int = syncSchemaVersion) {
        self.path = path
        self.title = title
        self.domain = domain
        self.vec_b64 = vec.toBase64()
        self.version = version
        let fmt = ISO8601DateFormatter()
        self.encoded_at = fmt.string(from: encodedAt)
        self.schema_version = schema_version
    }

    // Custom Decodable to tolerate legacy bundles without schema_version —
    // they decode as schema_version = 1 (the original wire version).
    private enum CodingKeys: String, CodingKey {
        case path, title, domain, vec_b64, version, encoded_at, schema_version
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.path = try c.decode(String.self, forKey: .path)
        self.title = (try? c.decode(String.self, forKey: .title)) ?? ""
        self.domain = (try? c.decode(String.self, forKey: .domain)) ?? "general"
        self.vec_b64 = try c.decode(String.self, forKey: .vec_b64)
        self.version = (try? c.decode(Int.self, forKey: .version)) ?? 1
        self.encoded_at = try? c.decode(String.self, forKey: .encoded_at)
        self.schema_version = (try? c.decode(Int.self, forKey: .schema_version)) ?? 1
    }

    public func toVector() -> Vector768 {
        Vector768.fromBase64(vec_b64)
    }
}

public enum SyncBundle {
    /// Serialize rows as JSONL text. UTF-8 safe. One trailing newline.
    public static func encode(_ rows: [SyncRow]) throws -> String {
        let encoder = JSONEncoder()
        // JSONL: one object per line. Ensure stable key order for diffability.
        encoder.outputFormatting = [.sortedKeys]
        var lines: [String] = []
        lines.reserveCapacity(rows.count)
        for row in rows {
            let data = try encoder.encode(row)
            guard let line = String(data: data, encoding: .utf8) else {
                throw SyncBundleError.encodingFailed
            }
            lines.append(line)
        }
        return lines.joined(separator: "\n") + "\n"
    }

    /// Parse JSONL text into rows. Blank lines skipped.
    public static func decode(_ text: String) throws -> [SyncRow] {
        let decoder = JSONDecoder()
        var out: [SyncRow] = []
        for rawLine in text.split(separator: "\n", omittingEmptySubsequences: true) {
            let line = rawLine.trimmingCharacters(in: .whitespaces)
            guard !line.isEmpty else { continue }
            guard let data = line.data(using: .utf8) else {
                throw SyncBundleError.decodingFailed
            }
            out.append(try decoder.decode(SyncRow.self, from: data))
        }
        return out
    }

    /// Write a bundle file atomically.
    public static func write(rows: [SyncRow], to url: URL) throws {
        let text = try encode(rows)
        try text.data(using: .utf8)?.write(to: url, options: .atomic)
    }

    /// Read a bundle file.
    public static func read(from url: URL) throws -> [SyncRow] {
        let data = try Data(contentsOf: url)
        guard let text = String(data: data, encoding: .utf8) else {
            throw SyncBundleError.decodingFailed
        }
        return try decode(text)
    }
}

public enum SyncBundleError: Error, Sendable {
    case encodingFailed
    case decodingFailed
}

File: src/tsuuid/sync_bundle.py
"""
tsuuid.sync_bundle — Cross-device row sync bundle format.

Each bundle file is JSON Lines (one row per line). Each row carries the
full metadata needed to upsert a vector on the receiving side:

    {"path": "...", "title": "...", "domain": "...",
     "vec_b64": "<base64 of float32 768 bytes>", "version": 1,
     "encoded_at": "2026-04-12T12:00:00Z"}

Why JSONL (not the raw SparseDelta bundle from tsuuid.bundle):
  - SparseDelta wire format has no path field; it's designed for intra-document
    version diffing with path implicit from context.
  - Cross-device sync routes new rows by path, so path must travel in-band.
  - JSONL is trivially cross-language (Swift Codable, Python stdlib) and
    remains debuggable with `cat`/`jq`.
"""

from __future__ import annotations

import base64
import json
from dataclasses import asdict, dataclass
from typing import Iterable, List, Optional


# Wire schema version. Bumped when SyncRow field layout changes in a way
# that older readers cannot safely ignore. Readers MUST refuse rows with
# schema_version > SYNC_SCHEMA_VERSION (forward-compat drift guard).
SYNC_SCHEMA_VERSION = 1


@dataclass
class SyncRow:
    path: str
    title: str
    domain: str
    vec_b64: str
    version: int = 1
    encoded_at: Optional[str] = None
    schema_version: int = SYNC_SCHEMA_VERSION

    def to_json(self) -> str:
        return json.dumps(asdict(self), separators=(",", ":"), ensure_ascii=False)

    @classmethod
    def from_json(cls, line: str) -> "SyncRow":
        obj = json.loads(line)
        return cls(
            path=obj["path"],
            title=obj.get("title", ""),
            domain=obj.get("domain", "general"),
            vec_b64=obj["vec_b64"],
            version=int(obj.get("version", 1)),
            encoded_at=obj.get("encoded_at"),
            schema_version=int(obj.get("schema_version", 1)),
        )

    def vec_bytes(self) -> bytes:
        return base64.b64decode(self.vec_b64)


def write_sync_bundle(rows: Iterable[SyncRow]) -> str:
    """Serialize rows as JSONL. One row per line, UTF-8 safe."""
    return "\n".join(r.to_json() for r in rows) + "\n"


def read_sync_bundle(text: str) -> List[SyncRow]:
    """Parse JSONL. Skips blank lines. Raises on malformed JSON."""
    rows: List[SyncRow] = []
    for line in text.splitlines():
        line = line.strip()
        if not line:
            continue
        rows.append(SyncRow.from_json(line))
    return rows
